sql: implement COPY ... TO STDOUT for CSV & TEXT

otan opened the pull request:
See individual commits for details.

These commits do *not* implement `COPY ... TO STDOUT BINARY`. That can be a later exercise!

fixes #85571

otan committed:
parser: implement parsing for COPY TO

This commit sets up the parsing required for COPY TO.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94408)
<!-- Reviewable:end -->

otan committed:
sql: refactor copy options

Strip out copy options to a function ready for re-use from the copy
machine in preparation for COPY TO.

Release note: None

rafiss reviewed (changes requested):
super cool! thanks for doing this!!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-:-NNKr0zi5OlFiBJiyKW9:bbph003)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[`-- commits` line 20 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NL3fKEv5hf-DFgISGcm):*
how does `FormatWireText` differ from `FmtPgwireText` or `fmtPgwireFormat`
___
*[`-- commits` line 48 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKqckU6JU3uRiXUEId):*
you could save the release note for this last one, and have it mention CSV support.
___
*[`pkg/cli/clisqlshell/sql.go` line 1878 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKqCfEBXHuiaVEHMZP:-NNKqCfEBXHuiaVEHMZQ:b-imkgok) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed2b5edd644585b98bdaeb36bb4d99db831e834f/pkg/cli/clisqlshell/sql.go#L1878)):*
> ```Go
> }
> 
> var copyToRe = regexp.MustCompile(`(?i)COPY.*TO\s+STDOUT`)
> ```

it seems possible that this could have a false positive match if COPY FROM is used with a table named "TO STDOUT" but maybe it doesn't matter..
___
*[`pkg/sql/pgwire/conn.go` line 1298 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NKswJdl7BKcfdyBFx1M:-NKswJdm0Ka39NdpdaoE:b72chz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5119a5625e61e1eb70a43ab01b50a87c8e98ed2d/pkg/sql/pgwire/conn.go#L1298)):*
> ```Go
> 	case tree.CopyIn, tree.CopyOut:
> 		// Nothing to do. The CommandComplete message has been sent elsewhere.
> 		panic(errors.AssertionFailedf("CopyIn statements should have been handled elsewhere " +
> ```

nit: update the error message
___
*[`pkg/sql/pgwire/conn.go` line 882 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKoFev0ZNnZ7T5HuPl:-NNKoFev0ZNnZ7T5HuPm:b-t8perj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2f72fda0f31da116f96c904697ffe3ef659626a/pkg/sql/pgwire/conn.go#L882)):*
> ```Go
> 		if cp, ok := stmts[i].AST.(*tree.CopyTo); ok {
> 			if len(stmts) != 1 {
> 				// NOTE(andrei): I don't know if Postgres supports receiving a COPY
> ```

does andrei's note apply to COPY TO? (the "taking over the connection" part)
___
*[`pkg/sql/sem/tree/copy.go` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NKswMq2DNGe-LGZEejX:-NKswMq2DNGe-LGZEejY:b7r62yb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5119a5625e61e1eb70a43ab01b50a87c8e98ed2d/pkg/sql/sem/tree/copy.go#L26)):*
> ```Go
> }
> 
> // CopyTo represents a COPY FROM statement.
> ```

nit: update the comment
___
*[`pkg/sql/copy/testdata/copy_to_binary` line 16 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKoaHtE7FEu44dQwzu:-NNKoaHtE7FEu44dQwzv:bx3k9fg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2f72fda0f31da116f96c904697ffe3ef659626a/pkg/sql/copy/testdata/copy_to_binary#L16)):*
> ```
> COPY t TO STDOUT BINARY
> ----
> ERROR: unimplemented: binary format for COPY TO not implemented (SQLSTATE 0A000)
> ```

does the issue link show up too?

is it because the issue is only in the hint? i wonder if we should just be putting them into the message so more people see them. (just wondering; that would be separate from this PR)
___
*[`pkg/sql/copy/testdata/copy_to_text` line 2 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKosab5laCzCmJV_9a:-NNKosac3CxD-uQ_ywBX:ba157l0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2f72fda0f31da116f96c904697ffe3ef659626a/pkg/sql/copy/testdata/copy_to_text#L2)):*
> ```
> exec-ddl
> CREATE TABLE t (id int primary key, t text);
> ```

i'd be curious to see how COPY TO handles weird stuff like the `bytea` type, arrays, tuples, time zones.
___
*[`pkg/sql/parser/testdata/copy` line 113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NKswahK4mDnrwG5oBtb:-NKswahK4mDnrwG5oBtc:b-v1438g) ([raw file](https://github.com/cockroachdb/cockroach/blob/5119a5625e61e1eb70a43ab01b50a87c8e98ed2d/pkg/sql/parser/testdata/copy#L113)):*
> ```
> COPY t TO 'file'
>           ^
> HINT: You have attempted to use a feature that is not yet implemented.
> ```

is there a way we can get an unimplementedWithIssue error for this?
___
*[`pkg/sql/copy_out_test.go` line 85 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKpAGG8u6WaNf5WNi4:-NNKpAGG8u6WaNf5WNi5:b1pdl31) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2f72fda0f31da116f96c904697ffe3ef659626a/pkg/sql/copy_out_test.go#L85)):*
> ```Go
> 
> // TestCopyOutRandom tests COPY TO on a set of random rows matches pgx.
> func TestCopyOutRandom(t *testing.T) {
> ```

cool test!
___
*[`pkg/sql/copy_out_test.go` line 99 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKpR2B-_pEHc3tG0Uq:-NNKpR2CAsCLF2ZIkuIM:b-ajoj6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2f72fda0f31da116f96c904697ffe3ef659626a/pkg/sql/copy_out_test.go#L99)):*
> ```Go
> 	colTypes := []*types.T{types.Int}
> 	colStr := "id INT PRIMARY KEY"
> 	for _, typ := range types.Scalar {
> ```

can non-scalar types be tested here too somehow?
___
*[`pkg/sql/pgwire/types_test.go` line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NL3fyLx9GD9S-SO2hlu:-NL3fyLx9GD9S-SO2hlv:b87ouxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/8258fb8c98c2ef444d28a5f6766d4faa7a346645/pkg/sql/pgwire/types_test.go#L44)):*
> ```Go
> // TestWriteTestDatumMatchesFormatWireText is required so long as writeTextDatum
> // has a separate implementation to tree.Datum.FormatWireText.
> func TestWriteTestDatumMatchesFormatWire(t *testing.T) {
> ```

should it be named `TestWriteTextDatumMatchesFormatWire` instead?
___
*[`pkg/sql/pgwire/types_test.go` line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NL3gBUB1VxUBfHWEvto:-NL3gBUB1VxUBfHWEvtp:b-n7c4sb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8258fb8c98c2ef444d28a5f6766d4faa7a346645/pkg/sql/pgwire/types_test.go#L50)):*
> ```Go
> 	defaultConv, defaultLoc := makeTestingConvCfg()
> 
> 	rng := rand.New(rand.NewSource(0))
> ```

nit: use `randutil.NewTestRand` instead


<!-- Sent from Reviewable.io -->

otan committed:
cli: allow the client to handle COPY TO

This commit enables the CLI to correctly route COPY TO commands to the
server.

Release note: None

otan committed:
sql: implement COPY ... TO STDOUT on the conn executor

Release note: None

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94408)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`-- commits` line 20 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NL3fKEv5hf-DFgISGcm):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

how does `FormatWireText` differ from `FmtPgwireText` or `fmtPgwireFormat`
</blockquote></details>

Done.
___
*[`-- commits` line 48 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKqckU6JU3uRiXUEId):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

you could save the release note for this last one, and have it mention CSV support.
</blockquote></details>

Done.
___
*[`pkg/cli/clisqlshell/sql.go` line 1878 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKqCfEBXHuiaVEHMZP:-NN_8-2-DzUBtoGS8iuy:b-g5opu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed2b5edd644585b98bdaeb36bb4d99db831e834f/pkg/cli/clisqlshell/sql.go#L1878)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it seems possible that this could have a false positive match if COPY FROM is used with a table named "TO STDOUT" but maybe it doesn't matter..
</blockquote></details>

yeah i didn't really see a good way around this without a proper parser :\
___
*[`pkg/sql/pgwire/conn.go` line 882 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKoFev0ZNnZ7T5HuPl:-NNZsuoI14_3JtQqD_3D:b-ydfe7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2f72fda0f31da116f96c904697ffe3ef659626a/pkg/sql/pgwire/conn.go#L882)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

does andrei's note apply to COPY TO? (the "taking over the connection" part)
</blockquote></details>

Unfortuntely so. We still take over the protocol here to avoid buffering too many rows at once. If we think that's okay, I propose adding that separaately.
___
*[`pkg/sql/copy/testdata/copy_to_binary` line 16 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKoaHtE7FEu44dQwzu:-NN_0V6u425hCbVcLhz7:b-4yo9pb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2f72fda0f31da116f96c904697ffe3ef659626a/pkg/sql/copy/testdata/copy_to_binary#L16)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

does the issue link show up too?

is it because the issue is only in the hint? i wonder if we should just be putting them into the message so more people see them. (just wondering; that would be separate from this PR)
</blockquote></details>

i'll add hints and details ehre.
___
*[`pkg/sql/copy/testdata/copy_to_text` line 2 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKosab5laCzCmJV_9a:-NN_0Rna3FADsrZvtj0H:bfv3uug) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2f72fda0f31da116f96c904697ffe3ef659626a/pkg/sql/copy/testdata/copy_to_text#L2)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i'd be curious to see how COPY TO handles weird stuff like the `bytea` type, arrays, tuples, time zones.
</blockquote></details>

added these tests (timezones already existed)
___
*[`pkg/sql/parser/testdata/copy` line 113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NKswahK4mDnrwG5oBtb:-NNZu0_q2pd__o9pMIXL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5119a5625e61e1eb70a43ab01b50a87c8e98ed2d/pkg/sql/parser/testdata/copy#L113)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is there a way we can get an unimplementedWithIssue error for this?
</blockquote></details>

Done.
___
*[`pkg/sql/pgwire/types_test.go` line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NL3fyLx9GD9S-SO2hlu:-NNZtu4FCS1tEXH0O741:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8258fb8c98c2ef444d28a5f6766d4faa7a346645/pkg/sql/pgwire/types_test.go#L44)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should it be named `TestWriteTextDatumMatchesFormatWire` instead?
</blockquote></details>

Done.
___
*[`pkg/sql/copy_out_test.go` line 85 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKpAGG8u6WaNf5WNi4:-NNZq3en4mZaKuHRh317:b-1bjplw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2f72fda0f31da116f96c904697ffe3ef659626a/pkg/sql/copy_out_test.go#L85)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

cool test!
</blockquote></details>

thanks :D
___
*[`pkg/sql/copy_out_test.go` line 99 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNKpR2B-_pEHc3tG0Uq:-NNZq31-3MqXBG8oRKh5:b5dpzqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2f72fda0f31da116f96c904697ffe3ef659626a/pkg/sql/copy_out_test.go#L99)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can non-scalar types be tested here too somehow?
</blockquote></details>

hmm, depends if there's something worth randomly testing here.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 1 of 3 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94408)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[`pkg/cli/clisqlshell/sql.go` line 1974 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNbhMSx8x7gur36_Z65:-NNbhMSy5IYYOizBRL0A:b-pge4e2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c04e78d6a341decfdfe973230fcfe92a94f01e72/pkg/cli/clisqlshell/sql.go#L1974)):*
> ```Go
> 		if scanner.FirstLexicalToken(c.concatLines) == lexbase.COPY {
> 			// Ideally this is parsed using the parser, but we've avoided doing so
> 			// for clisqlshell to be small.
> ```

Can you make this happen via `serverSideParse()` instead - let the server parse and tell you what you need to know. Your approach with a regexp is too brittle.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
heads up i split the first commit into #96583 as i believe that is cause for more scrutiny.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-:-NNdczV9B2s-rqkgMtqX:bxu8t1e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @rafiss)

___
*[`pkg/cli/clisqlshell/sql.go` line 1974 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNbhMSx8x7gur36_Z65:-NNdcl2C1kOs5Uh49rEi:bqhs9xh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c04e78d6a341decfdfe973230fcfe92a94f01e72/pkg/cli/clisqlshell/sql.go#L1974)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Can you make this happen via `serverSideParse()` instead - let the server parse and tell you what you need to know. Your approach with a regexp is too brittle.
</blockquote></details>

not sure how much it helps here:
```

root@127.0.0.1:26257/defaultdb> show syntax 'copy t to stdout';
  field |     message
--------+-------------------
  sql   | COPY t TO STDOUT
(1 row)

Time: 0ms total (execution 0ms / network 0ms)
```


<!-- Sent from Reviewable.io -->

otan committed:
sql: implement COPY TO with CSV format

This commit introduces the relevant CSV format components for COPY TO.

Release note (sql change): Implemented the `COPY ... TO STDOUT`
statement, which allows exporting a table or arbitrary query using the
postgres wire-compatible format. Only text & CSV format is supported.

rafiss reviewed (commented):
this seems great! thanks for doing it

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-:-NOKgtfKDuodwW519TgY:b2xci3e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> Reviewed 1 of 11 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94408)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/copy/testdata/copy_to_binary:
```diff
@@ -0,0 +1,18 @@
+----
+ERROR: unimplemented: binary format for COPY TO not implemented (SQLSTATE 0A000)
+HINT: You have attempted to use a feature that is not yet implemented.
+See: https://go.crdb.dev/issue-v/85571/v23.1
```
nit: would be nice to make separate issues for BINARY and `file` support.

rafiss replied:
filed:
https://github.com/cockroachdb/cockroach/issues/97180
https://github.com/cockroachdb/cockroach/issues/97181

otan committed:
sql: update error codes for certain COPY messages

Release note: None

otan commented:
bors r=rafiss

otan commented:
bors r=rafiss

otan commented:
bors is dying, flakes are alive

bors r=rafiss

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94408)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[`pkg/cli/clisqlshell/sql.go` line 1974 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/94408#-NNbhMSx8x7gur36_Z65:-NOue-SY0kpmFgU9fiIV:bz2rx1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c04e78d6a341decfdfe973230fcfe92a94f01e72/pkg/cli/clisqlshell/sql.go#L1974)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

not sure how much it helps here:
```

root@127.0.0.1:26257/defaultdb> show syntax 'copy t to stdout';
  field |     message
--------+-------------------
  sql   | COPY t TO STDOUT
(1 row)

Time: 0ms total (execution 0ms / network 0ms)
```
</blockquote></details>

For one, your thing does not handle string literals _nor SQL comments_.

Second, you're using `scanner.FirstLexicalToken` so you could actually decompose the tokens further and inspect them -- check that there are just 4 tokens, and the last 2 are "TO" and "STDOUT".

Third we could have a variant of SHOW SYNTAX that calls the (new) `parser.Tokens()` under the hood as that would do the tokenization server-side and give you the information you need.

This needs to be addressed -- we can't have a precedent of folk using regular expressions for SQL, as SQL is _not a regular language_.


<!-- Sent from Reviewable.io -->

knz commented:
filed #97508
